File: frontend/web/src/services/schedulingService.ts
import type { AppointmentSlot } from '../types/domain';
import { ApiError, apiRequest } from './api';

export interface AppointmentSlotResponse {
  id: string;
  veterinarianId: string;
  date: string;
  startTime: string;
  endTime: string;
  branch: string;
  status: string;
}

export interface AppointmentResponse {
  id: string;
  petId: string;
  veterinarianId: string;
  appointmentSlotId: string;
  scheduledStart: string;
  scheduledEnd: string;
  status: string;
  notes?: string;
  createdAt: string;
  updatedAt: string;
}

export interface CreateAppointmentRequest {
  petId: string;
  veterinarianId: string;
  appointmentSlotId: string;
  scheduledStart: string;
  scheduledEnd: string;
  notes?: string;
}

export interface UpdateAppointmentRequest {
  scheduledStart: string;
  scheduledEnd: string;
  notes?: string;
}

export interface ConflictCheckRequest {
  veterinarianId: string;
  scheduledStart: string;
  scheduledEnd: string;
  appointmentId?: string;
}

function toLocalSlot(slot: AppointmentSlotResponse): AppointmentSlot {
  return {
    id: slot.id,
    veterinarianId: slot.veterinarianId,
    veterinarianName: slot.veterinarianId,
    date: slot.date,
    startTime: slot.startTime,
    endTime: slot.endTime,
    branch: slot.branch,
    status: slot.status as AppointmentSlot['status'],
  };
}

export async function getAppointmentSlots(): Promise<AppointmentSlot[]> {
  const data = await apiRequest<AppointmentSlotResponse[]>('/appointments/available-slots');
  return data.map(toLocalSlot);
}

export async function getAppointmentById(id: string): Promise<AppointmentResponse | null> {
  try {
    return await apiRequest<AppointmentResponse>(`/appointments/${id}`);
  } catch (err) {
    if (err instanceof ApiError && err.status === 404) return null;
    throw err;
  }
}

export async function createAppointment(request: CreateAppointmentRequest): Promise<AppointmentResponse> {
  return await apiRequest<AppointmentResponse>('/appointments', { method: 'POST', body: JSON.stringify(request) });
}

export async function updateAppointment(id: string, request: UpdateAppointmentRequest): Promise<AppointmentResponse> {
  return await apiRequest<AppointmentResponse>(`/appointments/${id}`, { method: 'PUT', body: JSON.stringify(request) });
}

export async function cancelAppointment(id: string): Promise<void> {
  await apiRequest<void>(`/appointments/${id}`, { method: 'DELETE' });
}

export async function checkVetConflict(request: ConflictCheckRequest): Promise<boolean> {
  return await apiRequest<boolean>('/appointments/check-conflict', { method: 'POST', body: JSON.stringify(request) });
}

export function hasVetConflict(slots: AppointmentSlot[], candidate: Pick<AppointmentSlot, 'veterinarianId' | 'date' | 'startTime' | 'endTime'>): boolean {
  const toMinutes = (value: string) => {
    const [hour, minute] = value.split(':').map(Number);
    return hour * 60 + minute;
  };

  const start = toMinutes(candidate.startTime);
  const end = toMinutes(candidate.endTime);

  return slots.some((slot) => {
    if (slot.veterinarianId !== candidate.veterinarianId || slot.date !== candidate.date || slot.status === 'Cancelled') return false;
    const existingStart = toMinutes(slot.startTime);
    const existingEnd = toMinutes(slot.endTime);
    return start < existingEnd && end > existingStart;
  });
}

File: backend/api/src/PetCare.Application/Services/SchedulingService.cs
using FluentValidation;
using PetCare.Application.DTOs.Scheduling;
using PetCare.Application.Exceptions;
using PetCare.Application.Interfaces;
using PetCare.Domain.Entities;
using PetCare.Domain.Enums;

namespace PetCare.Application.Services;

/// <summary>
/// Scheduling business logic. This is the single source of truth for
/// appointment creation/update/cancellation rules and the veterinarian
/// overlap check, shared by every client (React, Flutter, ...) through
/// ASP.NET Core. See docs/database/scheduling-billing-approval-domain-model.md
/// for the underlying business rules.
/// </summary>
public class SchedulingService : ISchedulingService
{
    private readonly IAppointmentRepository _appointmentRepository;
    private readonly IAppointmentSlotRepository _appointmentSlotRepository;
    private readonly IVeterinarianRepository _veterinarianRepository;
    private readonly IUnitOfWork _unitOfWork;
    private readonly IValidator<CreateAppointmentRequest> _createValidator;
    private readonly IValidator<UpdateAppointmentRequest> _updateValidator;

    public SchedulingService(
        IAppointmentRepository appointmentRepository,
        IAppointmentSlotRepository appointmentSlotRepository,
        IVeterinarianRepository veterinarianRepository,
        IUnitOfWork unitOfWork,
        IValidator<CreateAppointmentRequest> createValidator,
        IValidator<UpdateAppointmentRequest> updateValidator)
    {
        _appointmentRepository = appointmentRepository;
        _appointmentSlotRepository = appointmentSlotRepository;
        _veterinarianRepository = veterinarianRepository;
        _unitOfWork = unitOfWork;
        _createValidator = createValidator;
        _updateValidator = updateValidator;
    }

    public async Task<IReadOnlyList<AppointmentResponse>> GetAppointmentsAsync(CancellationToken cancellationToken = default)
    {
        var appointments = await _appointmentRepository.GetAllAsync(cancellationToken);
        return appointments.Select(ToResponse).ToList();
    }

    public async Task<AppointmentResponse?> GetAppointmentByIdAsync(Guid id, CancellationToken cancellationToken = default)
    {
        var appointment = await _appointmentRepository.GetByIdAsync(id, cancellationToken);
        return appointment is null ? null : ToResponse(appointment);
    }

    public async Task<AppointmentResponse> CreateAppointmentAsync(CreateAppointmentRequest request, CancellationToken cancellationToken = default)
    {
        // Structural/referential validation (start < end, same-day, vet/slot exist).
        await _createValidator.ValidateAndThrowAsync(request, cancellationToken);

        // 1. Check veterinarian exists.
        var veterinarian = await _veterinarianRepository.GetByIdAsync(request.VeterinarianId, cancellationToken)
            ?? throw new NotFoundException($"Veterinarian '{request.VeterinarianId}' does not exist.");

        // 2. Check veterinarian is active.
        if (!veterinarian.Active)
        {
            throw new SchedulingConflictException($"Veterinarian '{request.VeterinarianId}' is not active.");
        }

        // 3. Check appointment slot exists.
        var slot = await _appointmentSlotRepository.GetByIdAsync(request.AppointmentSlotId, cancellationToken)
            ?? throw new NotFoundException($"Appointment slot '{request.AppointmentSlotId}' does not exist.");

        // 4. Check slot belongs to veterinarian.
        if (slot.VeterinarianId != request.VeterinarianId)
        {
            throw new SchedulingConflictException("The appointment slot does not belong to the specified veterinarian.");
        }

        var requestedDate = DateOnly.FromDateTime(request.ScheduledStart);
        var requestedStart = TimeOnly.FromDateTime(request.ScheduledStart);
        var requestedEnd = TimeOnly.FromDateTime(request.ScheduledEnd);

        // 5. Check requested time fits inside slot.
        if (requestedDate != slot.Date || requestedStart < slot.StartTime || requestedEnd > slot.EndTime)
        {
            throw new SchedulingConflictException("The requested appointment time does not fit inside the selected slot.");
        }

        // 6. Check no conflicting appointment exists.
        var hasConflict = await CheckConflictAsync(
            new ConflictCheckRequest
            {
                VeterinarianId = request.VeterinarianId,
                ScheduledStart = request.ScheduledStart,
                ScheduledEnd = request.ScheduledEnd
            },
            cancellationToken);

        if (hasConflict)
        {
            throw new SchedulingConflictException("The veterinarian already has a conflicting appointment for the requested time.");
        }

        // 7. Create appointment.
        var appointment = new Appointment
        {
            PetId = request.PetId,
            VeterinarianId = request.VeterinarianId,
            AppointmentSlotId = request.AppointmentSlotId,
            Date = requestedDate,
            StartTime = requestedStart,
            EndTime = requestedEnd,
            Status = AppointmentStatus.Reserved,
            Notes = request.Notes
        };

        slot.Status = AppointmentSlotStatus.Reserved;

        await _appointmentRepository.AddAsync(appointment, cancellationToken);

        // 8. Save changes (appointment insert + slot status update in one transaction/unit of work).
        await _unitOfWork.SaveChangesAsync(cancellationToken);

        return ToResponse(appointment);
    }

    public async Task<AppointmentResponse> UpdateAppointmentAsync(Guid id, UpdateAppointmentRequest request, CancellationToken cancellationToken = default)
    {
        await _updateValidator.ValidateAndThrowAsync(request, cancellationToken);

        var appointment = await _appointmentRepository.GetByIdAsync(id, cancellationToken)
            ?? throw new NotFoundException($"Appointment '{id}' does not exist.");

        var newDate = DateOnly.FromDateTime(request.ScheduledStart);
        var newStart = TimeOnly.FromDateTime(request.ScheduledStart);
        var newEnd = TimeOnly.FromDateTime(request.ScheduledEnd);

        var hasConflict = await CheckConflictAsync(
            new ConflictCheckRequest
            {
                VeterinarianId = appointment.VeterinarianId,
                ScheduledStart = request.ScheduledStart,
                ScheduledEnd = request.ScheduledEnd,
                AppointmentId = appointment.Id
            },
            cancellationToken);

        if (hasConflict)
        {
            throw new SchedulingConflictException("The veterinarian already has a conflicting appointment for the requested time.");
        }

        appointment.Date = newDate;
        appointment.StartTime = newStart;
        appointment.EndTime = newEnd;
        appointment.Notes = request.Notes;

        await _unitOfWork.SaveChangesAsync(cancellationToken);

        return ToResponse(appointment);
    }

    public async Task CancelAppointmentAsync(Guid id, CancellationToken cancellationToken = default)
    {
        var appointment = await _appointmentRepository.GetByIdAsync(id, cancellationToken)
            ?? throw new NotFoundException($"Appointment '{id}' does not exist.");

        appointment.Status = AppointmentStatus.Cancelled;

        // Cancelled appointments do not block a slot: free it for reuse.
        var slot = await _appointmentSlotRepository.GetByIdAsync(appointment.AppointmentSlotId, cancellationToken);
        if (slot is not null)
        {
            slot.Status = AppointmentSlotStatus.Available;
        }

        await _unitOfWork.SaveChangesAsync(cancellationToken);
    }

    public async Task<IReadOnlyList<AppointmentSlotResponse>> GetAvailableSlotsAsync(
        Guid? veterinarianId,
        DateOnly? date,
        CancellationToken cancellationToken = default)
    {
        var slots = await _appointmentSlotRepository.GetAvailableAsync(veterinarianId, date, cancellationToken);
        return slots.Select(ToSlotResponse).ToList();
    }

    public async Task<bool> CheckConflictAsync(ConflictCheckRequest request, CancellationToken cancellationToken = default)
    {
        var date = DateOnly.FromDateTime(request.ScheduledStart);
        var newStart = TimeOnly.FromDateTime(request.ScheduledStart);
        var newEnd = TimeOnly.FromDateTime(request.ScheduledEnd);

        var existingAppointments = await _appointmentRepository.GetActiveByVeterinarianAndDateAsync(
            request.VeterinarianId,
            date,
            cancellationToken);

        // Cancelled appointments do not block a slot; excluded defensively here
        // even though the repository is also expected to filter them out.
        return existingAppointments
            .Where(a => a.Id != request.AppointmentId && a.Status != AppointmentStatus.Cancelled)
            .Any(a => newStart < a.EndTime && newEnd > a.StartTime);
    }

    private static AppointmentResponse ToResponse(Appointment appointment) => new()
    {
        Id = appointment.Id,
        PetId = appointment.PetId,
        VeterinarianId = appointment.VeterinarianId,
        AppointmentSlotId = appointment.AppointmentSlotId,
        ScheduledStart = appointment.Date.ToDateTime(appointment.StartTime),
        ScheduledEnd = appointment.Date.ToDateTime(appointment.EndTime),
        Status = appointment.Status.ToString(),
        Notes = appointment.Notes,
        CreatedAt = appointment.CreatedAt,
        UpdatedAt = appointment.UpdatedAt
    };

    private static AppointmentSlotResponse ToSlotResponse(AppointmentSlot slot) => new()
    {
        Id = slot.Id,
        VeterinarianId = slot.VeterinarianId,
        Date = slot.Date,
        StartTime = slot.StartTime,
        EndTime = slot.EndTime,
        Branch = slot.Branch,
        Status = slot.Status.ToString()
    };
}
